### PR TITLE
[AMDGPU] Split Dpp8FI and Dpp16FI operands

### PR DESCRIPTION
**Pull request by jayfoad**
Split Dpp8FI and Dpp16FI into two different operands sharing an
AsmOperandClass. They are parsed and rendered identically as fi:1 but
the encoding is different: for DPP16 FI is a single bit, but for DPP8 it
uses two different special values in the src0 field. Having a dedicated
decoder for Dpp8FI allows it to reject other (non-special) src0 values
so that AMDGPUDisassembler::getInstruction no longer needs to call
isValidDPP8 to do post hoc validation of decoded DPP8 instructions.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/Disassembler/AMDGPUDisassembler.cpp [hunk 1]
```diff
@@ -474,13 +467,11 @@ DecodeStatus AMDGPUDisassembler::getInstruction(MCInst &MI, uint64_t &Size,
                         MI, DecW, Address, CS);
       if (Res && convertDPP8Inst(MI) == MCDisassembler::Success)
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/Disassembler/AMDGPUDisassembler.cpp:468]: Now `convertDPP8Inst` always returns success. I have patches which will clean this up some more.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td [hunk 2]
```diff
@@ -1092,7 +1092,20 @@ def DppBankMask : NamedIntOperand<i32, "bank_mask">;
...
+  let RenderMethod = "addImmOperands";
+  let IsOptional = true;
+  let DefaultMethod = "[this]() { return AMDGPUOperand::CreateImm(this, 0, SMLoc(), AMDGPUOperand::ImmTyDppFI); }";
+}
+let ParserMatchClass = DppFI, PrintMethod = "printDppFI" in {
+  def Dpp8FI : NamedIntOperand<i32, "fi"> {
+    let DecoderMethod = "decodeDpp8FI";
+  }
+  def Dpp16FI : NamedIntOperand<i32, "fi">;
+}
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.td:1108]: I'd like to know if there are ways to simplify these definitions.

**Review by rampitec — APPROVED** [review]
Nice!

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td (hunk 2 above)
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.td:1108]: TableGen treats identical instantiations as same records, so the operands implicitly generating identical `AsmOperandClass`es shouldn't be a problem.

With the `NamedIntOperand` class changed to support custom names, e.g.:
```
class NamedIntOperand<ValueType Type, string Prefix, bit Optional = 1,
                      string Name = NAME, string ConvertMethod = "nullptr">
    : CustomOperand<Type, Optional, Name> {
```

I think it should be possible to just do something like:
```
let DecoderMethod = "decodeDpp8FI" in
def Dpp8FI : NamedIntOperand<i32, "fi", 1, "DppFI">;
def Dpp16FI : NamedIntOperand<i32, "fi", 1, "DppFI">;
```

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.td]: Thank you! Done.

**Review by kosarev — APPROVED** [review]
Cool, thanks!

**Pull request merged**